File: packages/helm-runtime-sdk/node/src/transport.js
// HTTP transport for the helmstudio runtime SDK: fetch only, no Node built-ins,
// so the same file can serve a browser build once M6 decides browser access.

// The one typed error shape across languages (docs/design/04-packages.md §4).
const KINDS = {
  400: "Invalid", 413: "Invalid", 416: "Invalid", 422: "Invalid",
  401: "Unauthenticated",
  403: "Forbidden", 421: "Forbidden",
  404: "NotFound", 410: "NotFound",
  409: "Conflict",
  429: "QuotaExceeded", 507: "QuotaExceeded",
  501: "Unsupported",
  503: "Unavailable",
};

export class HelmError extends Error {
  constructor(status, code, message, details) {
    super(`${code} (${status}): ${message}`);
    this.status = status;
    this.code = code;
    this.details = details || {};
  }

  get kind() {
    if (this.status === 0) return "Unavailable";
    return KINDS[this.status] || "Internal";
  }
}

function value(v) {
  if (v instanceof Date) return v.toISOString();
  return String(v);
}

async function* events(response) {
  const reader = response.body.getReader();
  const decoder = new TextDecoder();
  let buffer = "";
  let ev = { id: "", name: "", data: [] };
  try {
    for (;;) {
      const { done, value: chunk } = await reader.read();
      if (done) return;
      buffer += decoder.decode(chunk, { stream: true });
      let nl;
      while ((nl = buffer.indexOf("\n")) >= 0) {
        const line = buffer.slice(0, nl).replace(/\r$/, "");
        buffer = buffer.slice(nl + 1);
        if (line === "") {
          if (ev.name || ev.data.length) {
            const data = ev.data.join("\n");
            yield { id: ev.id, name: ev.name, data, json: () => (data ? JSON.parse(data) : null) };
          }
          ev = { id: "", name: "", data: [] };
        } else if (line.startsWith(":")) {
          continue;
        } else if (line.startsWith("id:")) {
          ev.id = line.slice(3).trim();
        } else if (line.startsWith("event:")) {
          ev.name = line.slice(6).trim();
        } else if (line.startsWith("data:")) {
          ev.data.push(line.slice(5).replace(/^ /, ""));
        }
      }
    }
  } finally {
    reader.releaseLock();
  }
}

export class Transport {
  constructor(base, token, { fetch: fetchImpl } = {}) {
    this.base = base.replace(/\/+$/, "");
    this.token = token;
    this.fetch = fetchImpl || globalThis.fetch.bind(globalThis);
  }

  async request(method, path, { query = {}, headers = {}, expect, json, raw, contentType }) {
    const params = new URLSearchParams();
    for (const [k, v] of Object.entries(query)) {
      if (v === undefined || v === null) continue;
      if (Array.isArray(v)) v.forEach((x) => params.append(k, value(x)));
      else params.append(k, value(v));
    }
    const qs = params.toString();
    const h = { Authorization: `Bearer ${this.token}` };
    for (const [k, v] of Object.entries(headers)) {
      if (v !== undefined && v !== null) h[k] = value(v);
    }
    let body;
    if (json !== undefined) body = JSON.stringify(json);
    else if (raw !== undefined) body = raw;
    if (contentType) h["Content-Type"] = contentType;

    let response;
    try {
      response = await this.fetch(this.base + path + (qs ? `?${qs}` : ""), { method, headers: h, body });
    } catch (e) {
      throw new HelmError(0, "unavailable", String(e && e.message ? e.message : e));
    }
    if (!response.ok) {
      const text = await response.text();
      let doc;
      try {
        doc = JSON.parse(text);
      } catch {
        throw new HelmError(response.status, "unexpected_response", text);
      }
      throw new HelmError(response.status, doc.error || "unexpected_response", doc.message || "", doc.details);
    }
    if (expect === "sse") return events(response);
    if (expect === "raw") return response;
    if (expect === "json") return response.json();
    await response.arrayBuffer();
    return undefined;
  }
}

File: packages/helm-runtime-sdk/go/transport.go
package helm

import (
	"bufio"
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
)

type transport struct {
	base  string
	token string
	http  *http.Client
}

// do sends one request and returns the response when its status is 2xx. Any
// other status is read into *Error and the body closed.
func (t *transport) do(ctx context.Context, method, path string, query url.Values, header http.Header, body io.Reader, contentType string) (*http.Response, error) {
	u := t.base + path
	if len(query) > 0 {
		u += "?" + query.Encode()
	}
	req, err := http.NewRequestWithContext(ctx, method, u, body)
	if err != nil {
		return nil, fmt.Errorf("helm: building %s %s: %w", method, path, err)
	}
	for k, vs := range header {
		for _, v := range vs {
			req.Header.Add(k, v)
		}
	}
	if contentType != "" {
		req.Header.Set("Content-Type", contentType)
	}
	req.Header.Set("Authorization", "Bearer "+t.token)
	resp, err := t.http.Do(req)
	if err != nil {
		return nil, fmt.Errorf("helm: %s %s: %w", method, path, err)
	}
	if resp.StatusCode >= 200 && resp.StatusCode < 300 {
		return resp, nil
	}
	defer resp.Body.Close()
	raw, _ := io.ReadAll(io.LimitReader(resp.Body, 1<<20))
	e := &Error{Status: resp.StatusCode}
	if json.Unmarshal(raw, e) != nil || e.Code == "" {
		e.Code = "unexpected_response"
		e.Message = strings.TrimSpace(string(raw))
	}
	return nil, e
}

func encodeJSON(v any) (io.Reader, error) {
	b, err := json.Marshal(v)
	if err != nil {
		return nil, fmt.Errorf("helm: encoding the request body: %w", err)
	}
	return bytes.NewReader(b), nil
}

func decodeJSON[T any](resp *http.Response) (*T, error) {
	defer resp.Body.Close()
	var v T
	if err := json.NewDecoder(resp.Body).Decode(&v); err != nil {
		return nil, fmt.Errorf("helm: decoding the %s response: %w", resp.Request.URL.Path, err)
	}
	return &v, nil
}

func drain(resp *http.Response) error {
	_, _ = io.Copy(io.Discard, resp.Body)
	return resp.Body.Close()
}

// RawResponse is a byte stream: an asset's bytes or a thumbnail. Close it.
type RawResponse struct {
	Status        int
	ContentType   string
	ContentLength int64
	ContentRange  string
	ETag          string
	Body          io.ReadCloser
}

func newRawResponse(resp *http.Response) *RawResponse {
	return &RawResponse{Status: resp.StatusCode, ContentType: resp.Header.Get("Content-Type"),
		ContentLength: resp.ContentLength, ContentRange: resp.Header.Get("Content-Range"),
		ETag: resp.Header.Get("ETag"), Body: resp.Body}
}

// Event is one server-sent event.
type Event struct {
	ID   string
	Name string
	Data json.RawMessage
}

// Decode unmarshals the event's data, for example into an ItemEvent.
func (e Event) Decode(v any) error { return json.Unmarshal(e.Data, v) }

// EventStream reads server-sent events until the stream ends or Close.
type EventStream struct {
	body io.ReadCloser
	r    *bufio.Reader
}

func newEventStream(resp *http.Response) *EventStream {
	return &EventStream{body: resp.Body, r: bufio.NewReaderSize(resp.Body, 64<<10)}
}

// Next returns the next event, skipping comments. It returns io.EOF when the
// stream ends.
func (s *EventStream) Next() (Event, error) {
	var ev Event
	var data []string
	for {
		line, err := s.r.ReadString('\n')
		if err != nil && line == "" {
			if len(data) > 0 || ev.Name != "" {
				ev.Data = json.RawMessage(strings.Join(data, "\n"))
				return ev, nil
			}
			return Event{}, err
		}
		line = strings.TrimRight(line, "\r\n")
		switch {
		case line == "":
			if ev.Name != "" || len(data) > 0 {
				ev.Data = json.RawMessage(strings.Join(data, "\n"))
				return ev, nil
			}
		case strings.HasPrefix(line, ":"):
		case strings.HasPrefix(line, "id:"):
			ev.ID = strings.TrimSpace(line[3:])
		case strings.HasPrefix(line, "event:"):
			ev.Name = strings.TrimSpace(line[6:])
		case strings.HasPrefix(line, "data:"):
			data = append(data, strings.TrimPrefix(line[5:], " "))
		}
	}
}

// Close ends the stream.
func (s *EventStream) Close() error { return s.body.Close() }
